src/physics/cam/co2_cycle.F90

```diff
@@ -26,402 +24,312 @@ module co2_cycle
+
+      ! Local variables
+      integer :: m, mm
+      !----------------------------------------------------------------------------
+
+      if (.not. co2_flag) return
+
+      ! Add constituents and fluxes to history file
+      do m = 1, ncnst
+         mm = c_i(m)
+
+         call addfld(trim(cnst_name(mm))//'_BOT', horiz_only,  'A', 'kg/kg',   trim(cnst_longname(mm))//', Bottom Layer')
+         call addfld(cnst_name(mm),               (/ 'lev' /), 'A', 'kg/kg',   cnst_longname(mm))
+         call addfld(sflxnam(mm),                 horiz_only,  'A', 'kg/m2/s', trim(cnst_name(mm))//' surface flux')
+
+         call add_default(cnst_name(mm), 1, ' ')
+         call add_default(sflxnam(mm),   1, ' ')
+
+         ! The addfld call for the 'TM*' fields are made by default in the
+         ! constituent_burden module.
+         call add_default('TM'//trim(cnst_name(mm)), 1, ' ')
+      end do
+
+   end subroutine co2_init
+
+   !===============================================================================
+   subroutine co2_cycle_set_ptend(state, pbuf, ptend)
+
+      !-------------------------------------------------------------------------------
+      ! Purpose:
+      ! Set ptend, using aircraft CO2 emissions in ac_CO2 from pbuf
+      !-------------------------------------------------------------------------------
+
+      use physics_types,  only: physics_state, physics_ptend, physics_ptend_init
+      use physics_buffer, only: physics_buffer_desc, pbuf_get_index, pbuf_get_field
+      use constituents,   only: pcnst
+      use ppgrid,         only: pver
+      use physconst,      only: gravit
+
+      ! Arguments
+      type(physics_state), intent(in)    :: state
+      type(physics_buffer_desc), pointer :: pbuf(:)
+      type(physics_ptend), intent(out)   :: ptend     ! indivdual parameterization tendencies
+
+      ! Local variables
+      logical :: lq(pcnst)
+      integer :: ifld, ncol, k
+      real(r8), pointer :: ac_CO2(:,:)
+      !----------------------------------------------------------------------------
+
+      if (.not. co2_flag .or. .not. co2_readFlux_aircraft) then
+         call physics_ptend_init(ptend, state%psetcols, 'none')
+         return
+      end if
 
-   call physics_ptend_init(ptend, state%psetcols, 'co2_cycle_ac', lq=lq)
+      ! aircraft fluxes are added to 'CO2_FFF' and 'CO2' tendencies
+      lq(:)               = .false.
+      lq(co2_fff_glo_ind) = .true.
+      lq(co2_glo_ind)     = .true.
 
-   ifld = pbuf_get_index('ac_CO2')
-   call pbuf_get_field(pbuf, ifld, ac_CO2)
+      call physics_ptend_init(ptend, state%psetcols, 'co2_cycle_ac', lq=lq)
 
-   ! [ac_CO2] = 'kg m-2 s-1'
-   ! [ptend%q] = 'kg kg-1 s-1'
-   ncol = state%ncol
-   do k = 1, pver
-      ptend%q(:ncol,k,co2_fff_glo_ind) = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
-      ptend%q(:ncol,k,co2_glo_ind)     = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
-   end do
+      ifld = pbuf_get_index('ac_CO2')
```

Comment:
`ifld` should be a module variable with a more specific name (e.g., `idx_ac_CO2`) and initialized to `-1`. The `pbuf_get_index` call should be moved to `co2_init`.